packages/material-ui-system/src/index.d.ts

```diff
@@ -150,18 +150,18 @@ export const spacing: SimpleStyleFunction<
 export type SpacingProps = PropsFor<typeof spacing>;
 
 // style.js
-export interface StyleOptions<PropKey, Theme extends object> {
+export interface StyleOptions<PropKey, Theme extends object, PropType> {
   cssProperty?: PropKey | keyof CSS.Properties | false;
   prop: PropKey;
   /**
    * dot access in `Theme`
    */
   themeKey?: string;
-  transform?: (cssValue: unknown) => number | string;
+  transform?: (cssValue: PropType) => number | string | { [k: string]: any };
```

Comment:
I think at this point we can accept `any`.

Reply:
Sounds good. Not sure if returning false or so cause any problem anywhere, but if so it can be fixed then :)

Reply:
Returning `null` with ` cssProperty: false` probably results in runtime error when consumed via styled/makeStyles. It would result in style function returning null, which in turns ends with said consuming functions have a runtime error.

Reply:
Example: runtime error, when allowing return null: https://codesandbox.io/s/loving-hooks-kromb

Reply:
Alright then let's leave it like that. Maybe there already exists a type for possible inputs to jss? No idea where the boundary is where we pass those values.

Reply:
@eps1lon because of interop with other systems, like styled-components, it will be a bit hard to define a type, because then they would have to use this type too. Not that it is not possible to still do it, but there is simply no real contract there which other systems do. This bothered me a bit too until I came to this conclusion.

Reply:
> because of interop with other systems, like styled-components, it will be a bit hard to define a type, because then they would have to use this type too.

TypeScript does not use nominal typing. I would prefer a named type for documentation purposes and to be as wide as possible (does jss allow e.g booleans?).

Reply:
For documentation purposes, to be consistent, sure. Whether it can handle boolean, not sure. However the type can only be as wide as the intersection of types, the supported by CSS-in-JS libraries.

When I understood it correct (did not yet dig so far under the hood), JSS is only used when using styled from @material-ui/styles, right?

If so: even if JSS supports boolean, if styled-components does not, you can not add this type, because style should produce functions which works with all those supported libraries.

Reply:
Yeah just nitpicking for no real reason. This is perfectly fine.

Reply:
@eps1lon 
Just a note, sure nothing for this commit: this issue could be potentially solved, if you add an adaption for the style-result. Then as part of @material-ui/system, you will have function like `forStyledComponents` through which you can adapt the style functions:

```ts
import { styled } from 'styled-components';
import { forStyledComponents } from '@material-ui/system';

const StyledComponent = styled.div`
  ${forStyledComponents(spacing)}
`;
```

This is just an idea direction. There are solutions so the ease-of-use is not bad like in this simple example. Just don't want to post all the possibilities here, but might make a proposal for this later. This would enable you to type the own style function in a predictable way and then adapt result for the supported CSS-in-JS library-Libraries with a single function for each library you want to support.